apps/api/src/services/rewards.service.ts

```diff
@@ -0,0 +1,196 @@
+import keccak256 from "keccak256";
+import { encodeAbiParameters } from "viem";
+import { MerkleTree } from "merkletreejs";
+
+import { db } from "@/database/db.js";
+import {
+  getRewardsByEpoch,
+  getMerkleProof,
+  insertRewards,
+} from "@/database/repositories/rewards-repository.js";
+import { rewardsRoots, rewardsTree } from "@/database/schema/voting/defs.js";
+import { InsertReward } from "@/database/schema/voting/types.js";
+
+/**
+ * Service for handling reward-related operations
+ */
+export class RewardsService {
+  /**
+   * Calculate rewards for a given input
+   */
+  public async calculateRewards(): Promise<void> {
+    try {
+      const rewards = await this.calculate();
+      await insertRewards(rewards);
+    } catch (error) {
+      console.error("[RewardsService] Error in calculateRewards:", error);
+      throw error;
+    }
+  }
+
+  /**
+   * Allocates rewards for a given epoch by building a Merkle Tree and storing it in the database
+   * 
+   * This method:
+   * 1. Retrieves all rewards for the specified epoch
+   * 2. Builds a Merkle tree from the rewards (including a special "faux" leaf node)
+   * 3. Stores all tree nodes and the root hash in the database
+   * 4. Will eventually publish the root hash to the blockchain (TODO)
+   * 
+   * @param epochId The epoch ID (UUID) to allocate rewards for
+   * @throws Error if no rewards exist for the specified epoch
+   */
+  public async allocate(epochId: string): Promise<void> {
+    try {
+      const rewards = await getRewardsByEpoch(epochId);
+
+      // Build Merkle tree if we have rewards
+      if (rewards.length == 0) {
+        throw Error("no rewards to allocate");
+      }
+
+      // Prepend a faux leaf
+      const leaves = [createFauxLeafNode(epochId), ...rewards.map((reward) => reward.leafHash)];
+      const merkleTree = new MerkleTree(leaves, keccak256, {
+        sortPairs: true,
+        hashLeaves: false,
+        sortLeaves: true,
+      });
+
+      const layers = merkleTree.getLayers();
+      const treeNodes: {
+        id: string;
+        epoch: string;
+        level: number;
+        idx: number;
+        hash: Uint8Array;
+      }[] = [];
+
+      for (let level = 0; level < layers.length; level++) {
+        const layer = layers[level];
+
+        if (layer) {
+          for (let idx = 0; idx < layer.length; idx++) {
+            const hash = layer[idx];
+
+            if (hash) {
+              treeNodes.push({
+                id: crypto.randomUUID(),
+                epoch: epochId,
+                level,
+                idx,
+                hash: new Uint8Array(hash),
+              });
+            }
+          }
+        }
+      }
+
+      const rootHash = merkleTree.getRoot();
+
+      await db.transaction(async (tx) => {
+        //TODO: call addDistribution blockchain method
```

Comment:
Here we call the smart contract method [addAllocation](https://docs.google.com/document/d/1tA3DknAQqiCIFkeRjPuZxPBXLYZqqSNJ9q774p3CQbg/edit?tab=t.0#heading=h.l9kfj6etlu8x)